source/I2C.c

```
@@ -0,0 +1,632 @@
+    uint16_t timeout = 500;
+
+    while (__HAL_I2C_GET_FLAG(handle, I2C_FLAG_BTF) == RESET)
+    {
+        /* Check for NACK */
+        if (__HAL_I2C_GET_FLAG(handle, I2C_FLAG_AF) == SET)
+        {
+            __HAL_I2C_CLEAR_FLAG(handle, I2C_FLAG_AF);
+            return I2C_ERROR;
+        }
+
+        if (count >= timeout)
+        {
+            return I2C_ERROR;
+        }
+        count++;
+        vTaskDelay(50);
+    }
+    return I2C_OK;
+}
+
+static KI2CStatus hal_i2c_check_txe_timeout(I2C_HandleTypeDef * handle)
+{
+    uint16_t count = 0;
+    uint16_t timeout = 500;
+
+    while (__HAL_I2C_GET_FLAG(handle, I2C_FLAG_TXE) == RESET)
+    {
+        /* Check for NACK */
+        if (__HAL_I2C_GET_FLAG(handle, I2C_FLAG_AF) == SET)
+        {
+            __HAL_I2C_CLEAR_FLAG(handle, I2C_FLAG_AF);
+            return I2C_ERROR;
+        }
+        if (count >= timeout)
+        {
+            return I2C_ERROR;
+        }
+        count++;
+        vTaskDelay(50);
+    }
+    return I2C_OK;
+}
+
+static KI2CStatus hal_i2c_master_request_read(I2C_HandleTypeDef * hal_handle, uint16_t addr)
+{
+    /* Enable Acknowledge */
+    hal_handle->Instance->CR1 |= I2C_CR1_ACK;
+
+    /* Generate Start */
+    hal_handle->Instance->CR1 |= I2C_CR1_START;
+
+    /* Wait for SB */
+    if (hal_i2c_check_flag_timeout(hal_handle, I2C_FLAG_SB, RESET) != I2C_OK)
+    {
+        return I2C_ERROR;
+    }
+
+    /* Send slave address */
+    hal_handle->Instance->DR = I2C_7BIT_ADD_READ(addr);
+
+    /* Wait for ADDR */
+    if (hal_i2c_check_addr_timeout(hal_handle, I2C_FLAG_ADDR) != I2C_OK)
+    {
+        return I2C_ERROR;
+    }
+
+    return I2C_OK;
+}
+
+static KI2CStatus hal_i2c_master_setup_read(I2C_HandleTypeDef * hal_handle, uint16_t addr, int len)
+{
+    // I2C_7BIT_ADD_READ expects an address already shifted
+    uint16_t slave_addr = addr << 1;
+
+    /* Check if I2C Busy */
+    if (hal_i2c_check_flag_timeout(hal_handle, I2C_FLAG_BUSY, SET) != I2C_OK)
```

Do you want to add more kinds of errors i.e I2C_ERROR_BUSY to `KI2CStatus`? if we did, we could just return the error code directly here instead of blindly returning `I2C_ERROR`, and bubble some of these specific errors up to the end-user API